src/microbe_stage/MicrobeAI.cs

```diff
@@ -370,43 +384,86 @@ private void RunAndTumble(Random random)
-        }
+            // Note this is about absorbed quantities (which is all microbe has access to), not real ones
+            float compoundDifference = 0.0f;
 
-        float compoundDifference = 0.0f;
-        foreach (var compound in usefulCompounds)
-        {
-            compoundDifference += compound.Value;
+            // No need to add a difference if compound was not absorbed
+            if (microbe.TotalAbsorbedCompounds.ContainsKey(compoundWeight.Key))
+                compoundDifference += microbe.TotalAbsorbedCompounds[compoundWeight.Key];
+
+            // Idem if not absorbed before
+            if (previouslyAbsorbedCompounds.ContainsKey(compoundWeight.Key))
+                compoundDifference -= previouslyAbsorbedCompounds[compoundWeight.Key];
+
+            compoundDifference *= compoundWeight.Value;
+            gradientValue += compoundDifference;
         }
 
+        // Implement a detection threshold to possibly rule out too tiny variations
+        float differenceDetectionThreshold = 0;
+
         // If food density is going down, back up and see if there's some more
-        if (compoundDifference < 0 && random.Next(0, 10) < 9)
+        if (gradientValue < -differenceDetectionThreshold && random.Next(0, 10) < 9)
         {
             MoveWithRandomTurn(2.5f, 3.0f, random);
         }
 
         // If there isn't any food here, it's a good idea to keep moving
-        if (compoundDifference == 0 && random.Next(0, 10) < 5)
+        if (Math.Abs(gradientValue) <= differenceDetectionThreshold && random.Next(0, 10) < 5)
         {
             MoveWithRandomTurn(0.0f, 0.4f, random);
         }
 
-        // If positive last step you gained compounds, so let's stick around
-        if (compoundDifference > 0)
+        // If positive last step you gained compounds, so let's move toward the source
+        if (gradientValue > differenceDetectionThreshold)
         {
-            // There's a decent chance to turn most of the way around
+            // There's a decent chance to turn by 90° to explore gradient
+            // 180° is useless since previous position let you absorb less compounds already
             if (random.Next(0, 10) < 4)
             {
-                MoveWithRandomTurn(0.0f, 3.0f, random);
+                MoveWithRandomTurn(0.0f, 1.5f, random);
             }
         }
     }
 
+    /// <summary>
+    ///   Prioritizing compounds that are stored in lesser quantities.
+    ///   If ATP-producing compounds are low (less than half storage capacities),
+    ///   non ATP-related compounds are discarded.
+    /// </summary>
+    /// <returns> A dictionary of weights for each (useful) compound, to be used in search algorithms.</returns>
+    private Dictionary<Compound, float> ComputeCompoundsSearchWeights(
+        List<KeyValuePair<Compound, float>> usefulCompounds)
+    {
+        // Vital compounds are *direct* ATP producers
```

Comment:
I think a TODO comment should be added here to explain that this is a *slight* shortcut, as in theory we don't know here if iron is truly useful  for ATP production or if some future process uses it as input but doesn't make ATP.
The logic is fine enough now but I want that clarifying comment.

Reply:
Indeed. I also figured out that the code was a bit weak if further compounds were added (e.g. a third component for reproduction would still be searched, as such). I thought of having some kind of categories (Vital, ATP-related, reproduction...), which would allow to check if a Compound var is part of it or not. That'd rather be refactoring work though I think, so I kept it a bit aside for the moment. 

These do not cover multiple usages anyways.